sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/CollectionSizeAndArrayLengthCheck.kt

```diff
@@ -0,0 +1,81 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtBinaryExpression
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.resolve.calls.util.getResolvedCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.predictRuntimeIntValue
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+val COLLECTION_SIZE_METHOD = FunMatcher(qualifier = "kotlin.collections.Collection", name = "size") {
+    withNoArguments()
+}
+val ARRAY_SIZE_METHOD = FunMatcher(qualifier = "kotlin.Array", name = "size") {
+    withNoArguments()
+}
+
+val MATCHERS = listOf(COLLECTION_SIZE_METHOD, ARRAY_SIZE_METHOD)
+
+val ISSUE_MESSAGE_SIZE_NEVER_LT = """The size of an array/collection is never "<0", update this test to use ".isEmpty()"."""
+val ISSUE_MESSAGE_SIZE_ALWAYS_GTEQ =
+    """The size of an array/collection is always ">=0", update this test to either ".isNotEmpty()" or ".isEmpty()"."""
+
+@Rule(key = "S3981")
+class CollectionSizeAndArrayLengthCheck : AbstractCheck() {
+
+    override fun visitBinaryExpression(bet: KtBinaryExpression, fileCtx: KotlinFileContext) {
+        val ctx = fileCtx.bindingContext
+        val opToken = bet.operationToken
+
+        val leftIntValue = bet.left?.predictRuntimeIntValue(ctx)
+        val rightIntValue = bet.right?.predictRuntimeIntValue(ctx)
+        if (!leftIntValue.isZeroOrNegative() && !rightIntValue.isZeroOrNegative()) return
+
+        val isLeftZero = leftIntValue.isZeroOrNegative()
+
+        val testedExpr = if (isLeftZero) bet.right else bet.left
+        val integerValue = if (isLeftZero) leftIntValue else rightIntValue
+        val opWithEq = if (isLeftZero) KtTokens.LTEQ else KtTokens.GTEQ
+        val opWithoutEq = if (isLeftZero) KtTokens.GT else KtTokens.LT
+
+        if (testedExpr is KtDotQualifiedExpression && MATCHERS.any { it.matches(testedExpr.getResolvedCall(ctx)) }) {
+            if (opToken == opWithEq) {
+                fileCtx.reportIssue(bet, ISSUE_MESSAGE_SIZE_ALWAYS_GTEQ)
+            } else if (opToken == opWithoutEq || (integerValue.isNegative() && opToken == KtTokens.EQEQ)) {
+                fileCtx.reportIssue(bet, ISSUE_MESSAGE_SIZE_NEVER_LT)
+            }
+        }
+
+    }
```

Comment:
I find this a little confusing because there is a lot happening that is, in fact, all closely related. We first have a check about whether both sides are <=0, then we assign the fact that left is <=0 to a val, then we use that val to perform 4 different if-else operations. In fact, there are two different execution streams but the way it currently is written means they are extremely interleaved with each other. I think here we can take a few more lines of code to be more explicit about these execution flows. For instance, something along these lines: 
```suggestion

        if (leftIntValue.isZeroOrNegative()) {
            checkConditionsAndSelectMessage(ctx, opToken, bet.right, leftIntValue, KtTokens.LTEQ, KtTokens.GT)
        } else if (rightIntValue.isZeroOrNegative()) {
            checkConditionsAndSelectMessage(ctx, opToken, bet.left, rightIntValue, KtTokens.GTEQ, KtTokens.LT)
        } else {
            return
        }?.let {
            fileCtx.reportIssue(bet, it)
        }
    }

    private fun checkConditionsAndSelectMessage(
        ctx: BindingContext,
        opToken: IElementType,
        testedExpr: KtExpression?,
        integerValue: Int?,
        opWithEq: KtSingleValueToken,
        opWithoutEq: KtSingleValueToken,
    ): String? {
        if (testedExpr is KtDotQualifiedExpression && MATCHERS.any { it.matches(testedExpr.getResolvedCall(ctx)) }) {
            if (opToken == opWithEq) {
                return ISSUE_MESSAGE_SIZE_ALWAYS_GTEQ
            } else if (opToken == opWithoutEq || (integerValue.isNegative() && opToken == KtTokens.EQEQ)) {
                return ISSUE_MESSAGE_SIZE_NEVER_LT
            }
        }
        return null
    }
```